docs/08-gitcoin-identity-staking-management.md

```diff
@@ -0,0 +1,105 @@
+mapping(address => uint256[]) selfStakes;
+
+// mapping(stakerAddress => mapping(stakeeAddress => uint[])) communityStakes;
+mapping(address => mapping(address => uint256[])) communityStakes;
+
+mapping(uint256 => Stake) stakes;
+
+// mapping(stakerAddress => stakeeAddress[]) addressesStakedOnByUser;
+mapping(address => address[]) addressesStakedOnByUser;
+
+// mapping(stakeeAddress => stakerAddress[]) addressesStakingOnUser;
+mapping(address => address[]) addressesStakingOnUser;
+
+struct Stake {
+  uint256 amount;
+  uint256 unlockTime;
+}
+```
+
+We considered using an enumerable set instead of lists to track Stake IDs. Although all operations besides writing were more efficient with an enumerable set, the difference when writing a new value/staking was considerably different, and assuming that staking would be the most common operation, we decided to use lists instead. Here is a comparison of the gas costs for each operation: https://miro.com/app/board/uXjVNFnedyI=/
+
+## Methods
+
+Relevant methods and their responsibilities
+
+```solidity
+function selfStake(address staker, Stake stake) external
+```
+
+- Transfers tokens from the staker to the contract.
+- Creates a new `Stake` and adds it to the `stakes` mapping.
+- Adds the `Stake` ID to the `selfStakes` mapping for the user.
+
+```solidity
+function communityStake(address staker, address stakee, Stake stake) external
+```
+
+- Transfers tokens from the staker to the contract.
+- Creates a new `Stake` and adds it to the `stakes` mapping.
+- Maps the stakee and newly created Stake to the staker
+- Maps stakee and staker accordingly for `addressesStakedOnByUser` and `addressesStakingOnUser` these will be used for slashing
+  **Note**: Since this function involves updating 3 arrays, it might be worth while testing enumerable sets for this function.
+
+```solidity
+function unStake(uint256 stakeId) external
+```
+
+- Checks that the stake is unlocked.
+- Checks that address controls the stake.
+- Removes the stake from the `selfStakes` mapping for the user.
+- Removes the stake from the `stakes` mapping.
+- Transfers tokens from the contract to the staker.
+
+```solidity
+function communityUnStake(uint256 stakeId) external
+```
+
+- Checks that the stake is unlocked.
+- Checks that address controls the stake.
+- Removes the stake from the `communityStakes` mapping for the user. (Implementation)[https://github.com/OpenZeppelin/openzeppelin-contracts/blob/cffb2f1ddcd87efd68effc92cfd336c5145acabd/contracts/utils/structs/EnumerableSet.sol#L93] of efficient removal from array.
+- Removes the stake from the `stakes` mapping.
+- Transfers tokens from the contract to the staker.
+- Removes stakee and staker accordingly for `addressesStakedOnByUser` and `addressesStakingOnUser` these will be used for slashing
+  **Note**: Since this function involves updating 3 arrays, it might be worth while testing enumerable sets for this function.
+
+```solidity
+function reStake(uint256 stakeId) external
```

Comment:
Shall we add a duration parameter here?
We can pass in the same value after all ...

Also, what will be the starting time for the restake?
a. the `unstakeTime`
b. or the `block.timestamp`

Update: I just saw in the last paragraph of this doc, that it's option B ...